src/App.tsx

```diff
@@ -1,26 +1,181 @@
-/* eslint-disable jsx-a11y/control-has-associated-label */
-import React from 'react';
+import React, { useCallback, useEffect, useRef, useState } from 'react';
+import { USER_ID, getTodos, deleteTodo, addTodo } from './api/todos';
+import { TodoList } from './components/TodoList';
+import { Todo } from './types/Todo';
+import { Error } from './types/Error';
+import { Filter } from './types/Filter';
+import { getFilteredTodos } from './helpers';
+import { TodoHeader } from './components/TodoHeader';
+import { TodoFooter } from './components/TodoFooter';
+import { ErrorNotification } from './components/ErrorNotification';
 import { UserWarning } from './UserWarning';
-
-const USER_ID = 0;
+import { TodoItem } from './components/TodoItem';
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [currentError, setCurrentError] = useState<Error | null>(null);
+  const [currentFilter, setCurrentFilter] = useState<Filter>(Filter.All);
+  const [deleteTodoId, setDeleteTodoId] = useState<Todo['id'] | null>(null);
+  const [addTodoTitle, setAddTodoTitle] = useState<string>('');
+  const [tempTodo, setTempTodo] = useState<null | Todo>(null);
+  const [isNewTodoLoading, setIsNewTodoLoading] = useState<boolean>(false);
+
+  const hasCompletedTodos = todos.some((todo: Todo) => todo.completed);
+
+  const activeTodosAmount = todos.filter(
+    (todo: Todo) => !todo.completed,
+  ).length;
+
+  const newTodoInput = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    setCurrentError(null);
+
+    getTodos()
+      .then(setTodos)
+      .catch(() => {
+        setCurrentError(Error.CannotLoad);
+      });
+  }, []);
+
+  useEffect(() => {
+    if (currentError) {
+      newTodoInput.current?.focus();
+
+      setTimeout(() => setCurrentError(null), 3000);
+    }
+  }, [currentError]);
+
+  useEffect(() => {
+    newTodoInput.current?.focus();
+  }, [todos, newTodoInput]);
+
+  useEffect(() => {
+    if (deleteTodoId) {
+      setCurrentError(null);
+
+      deleteTodo(deleteTodoId)
+        .then(() => {
+          getTodos()
+            .then(setTodos)
+            .catch(() => {
+              setCurrentError(Error.CannotLoad);
+            });
+        })
+        .catch(() => {
+          setCurrentError(Error.CannotDelete);
+        })
+        .finally(() => setDeleteTodoId(null));
+    }
+  }, [deleteTodoId]);
+
+  const deleteTodoById = useCallback((id: number) => {
+    setDeleteTodoId(id);
+
+    deleteTodo(id)
+      .then(() =>
+        setTodos((currentTodos: Todo[]) =>
+          currentTodos.filter((todo: Todo) => todo.id !== id),
+        ),
+      )
+      .catch(() => {
+        setCurrentError(Error.CannotDelete);
+      })
+      .finally(() => setDeleteTodoId(null));
+  }, []);
+
+  const createTodo = useCallback(() => {
+    const todoTitle = addTodoTitle?.trim();
+
+    if (!todoTitle.length) {
+      setCurrentError(Error.EmptyTitle);
+
+      return;
+    }
+
+    setTempTodo({
```

Comment:
The idea of having a tempTodo is working, but don't you find it overcomplicated? 
Like, you have a separate state for it and exception in your jsx to display that one item. 
And all the effort - just to display an item with loader. 

Wouldn't it be easier to add a temporary item in the list before the callback, like:

` setTodos(current => [...current, tempTodo])`

trigger the API call
then, on success you can remove your temporary item with the real one(received from the server)
or, on fail, catch the error and remove temp item from the list.

Reply:
In the readme it is specified that there has to be state for the tempTodo and to NOT add it to the todos array.
However if you want of course I can try to do it this way:)